src/libsystemd/sd-journal/journal-file.c

```diff
@@ -2775,6 +2775,59 @@ enum {
         TEST_RIGHT
 };
 
+static int generic_array_bisect_one(
+                JournalFile *f,
+                uint64_t a, /* offset of entry array object. */
+                uint64_t i, /* index of the entry item we will test. */
+                uint64_t needle,
+                int (*test_object)(JournalFile *f, uint64_t p, uint64_t needle),
+                direction_t direction,
+                uint64_t *left,
+                uint64_t *right,
+                uint64_t *ret_offset) {
+
+        Object *array;
+        uint64_t p;
+        int r;
+
+        assert(f);
+        assert(test_object);
+        assert(left);
+        assert(right);
+        assert(*left <= i);
+        assert(i <= *right);
+
+        r = journal_file_move_to_object(f, OBJECT_ENTRY_ARRAY, a, &array);
+        if (r < 0)
+                return r;
+
+        p = journal_file_entry_array_item(f, array, i);
+        if (p <= 0)
+                r = -EBADMSG;
+        else
+                r = test_object(f, p, needle);
+        if (IN_SET(r, -EBADMSG, -EADDRNOTAVAIL)) {
+                log_debug_errno(r, "Encountered invalid entry while bisecting, cutting algorithm short.");
+                *right = i;
+                return -ENOANO; /* recognizable error */
```

Comment:
And here I thought that nothing would beat `ENOTNAM` when it comes to cryptic errnos, nice :)